File: frontend/public/py_src/service/profile_detection.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class ProfileDetectionResult:
    profile: str
    reason: str
    has_vrm0_extension: bool
    has_vrm1_extension: bool
    has_spring_extension: bool
    generator: str

    def as_dict(self) -> dict[str, Any]:
        return {
            "profile": self.profile,
            "reason": self.reason,
            "has_vrm0_extension": self.has_vrm0_extension,
            "has_vrm1_extension": self.has_vrm1_extension,
            "has_spring_extension": self.has_spring_extension,
            "generator": self.generator,
        }


def _as_mapping(value: Any) -> dict[str, Any]:
    return value if isinstance(value, dict) else {}


def detect_profile(
    json_data: dict[str, Any] | None,
    source_path: str = "",
) -> ProfileDetectionResult:
    data = _as_mapping(json_data)
    extensions = _as_mapping(data.get("extensions"))
    asset = _as_mapping(data.get("asset"))

    has_vrm0 = "VRM" in extensions
    has_vrm1 = "VRMC_vrm" in extensions
    has_spring = "VRMC_springBone" in extensions
    generator = str(asset.get("generator") or "")

    normalized_path = source_path.lower()
    normalized_generator = generator.lower()
    is_vroid_hint = "vroid" in normalized_path or "vroid" in normalized_generator

    if is_vroid_hint and (has_vrm0 or has_vrm1):
        return ProfileDetectionResult(
            profile="vroid",
            reason="vroid hint detected in path/generator",
            has_vrm0_extension=has_vrm0,
            has_vrm1_extension=has_vrm1,
            has_spring_extension=has_spring,
            generator=generator,
        )

    if has_vrm1:
        return ProfileDetectionResult(
            profile="generic",
            reason="VRMC_vrm extension detected",
            has_vrm0_extension=has_vrm0,
            has_vrm1_extension=has_vrm1,
            has_spring_extension=has_spring,
            generator=generator,
        )

    if has_vrm0:
        return ProfileDetectionResult(
            profile="generic",
            reason="VRM extension detected",
            has_vrm0_extension=has_vrm0,
            has_vrm1_extension=has_vrm1,
            has_spring_extension=has_spring,
            generator=generator,
        )

    return ProfileDetectionResult(
        profile="generic",
        reason="unknown metadata, fallback to generic",
        has_vrm0_extension=has_vrm0,
        has_vrm1_extension=has_vrm1,
        has_spring_extension=has_spring,
        generator=generator,
    )

File: frontend/src/features/convert/services/profileDetection.ts
export type ProfileDetectionResult = {
  profile: "generic" | "vroid";
  reason: string;
  hasVrm0Extension: boolean;
  hasVrm1Extension: boolean;
  hasSpringExtension: boolean;
  generator: string;
};

const GLB_MAGIC = 0x46546c67;
const JSON_CHUNK_TYPE = 0x4e4f534a;

function asRecord(value: unknown): Record<string, unknown> {
  return value && typeof value === "object" && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : {};
}

function detectProfileFromJson(
  jsonData: Record<string, unknown> | null,
  sourcePath = "",
): ProfileDetectionResult {
  const data = asRecord(jsonData);
  const extensions = asRecord(data.extensions);
  const asset = asRecord(data.asset);

  const hasVrm0 = "VRM" in extensions;
  const hasVrm1 = "VRMC_vrm" in extensions;
  const hasSpring = "VRMC_springBone" in extensions;
  const generator = typeof asset.generator === "string" ? asset.generator : "";

  const normalizedPath = sourcePath.toLowerCase();
  const normalizedGenerator = generator.toLowerCase();
  const isVroidHint =
    normalizedPath.includes("vroid") || normalizedGenerator.includes("vroid");

  if (isVroidHint && (hasVrm0 || hasVrm1)) {
    return {
      profile: "vroid",
      reason: "vroid hint detected in file name/generator",
      hasVrm0Extension: hasVrm0,
      hasVrm1Extension: hasVrm1,
      hasSpringExtension: hasSpring,
      generator,
    };
  }

  if (hasVrm1) {
    return {
      profile: "generic",
      reason: "VRMC_vrm extension detected",
      hasVrm0Extension: hasVrm0,
      hasVrm1Extension: hasVrm1,
      hasSpringExtension: hasSpring,
      generator,
    };
  }

  if (hasVrm0) {
    return {
      profile: "generic",
      reason: "VRM extension detected",
      hasVrm0Extension: hasVrm0,
      hasVrm1Extension: hasVrm1,
      hasSpringExtension: hasSpring,
      generator,
    };
  }

  return {
    profile: "generic",
    reason: "unknown metadata, fallback to generic",
    hasVrm0Extension: hasVrm0,
    hasVrm1Extension: hasVrm1,
    hasSpringExtension: hasSpring,
    generator,
  };
}

function parseGlbJson(buffer: ArrayBuffer): Record<string, unknown> | null {
  if (buffer.byteLength < 20) {
    return null;
  }

  const view = new DataView(buffer);
  if (view.getUint32(0, true) !== GLB_MAGIC) {
    return null;
  }

  let offset = 12;
  while (offset + 8 <= buffer.byteLength) {
    const chunkLength = view.getUint32(offset, true);
    const chunkType = view.getUint32(offset + 4, true);
    offset += 8;

    if (offset + chunkLength > buffer.byteLength) {
      return null;
    }

    if (chunkType === JSON_CHUNK_TYPE) {
      const chunk = new Uint8Array(buffer, offset, chunkLength);
      const jsonText = new TextDecoder().decode(chunk).replace(/\u0000+$/u, "");
      const parsed = JSON.parse(jsonText) as unknown;
      return asRecord(parsed);
    }

    offset += chunkLength;
  }

  return null;
}

export async function detectProfileFromFile(
  file: File,
): Promise<ProfileDetectionResult | null> {
  try {
    const buffer = await file.arrayBuffer();
    const jsonData = parseGlbJson(buffer);
    if (!jsonData) {
      return null;
    }

    return detectProfileFromJson(jsonData, file.name);
  } catch {
    return null;
  }
}
